cpp/src/dictionary/detail/concatenate.cu

```diff
@@ -0,0 +1,268 @@
+ */
+struct dispatch_compute_indices {
+  template <typename Element>
+  typename std::enable_if_t<cudf::is_relationally_comparable<Element, Element>(),
+                            std::unique_ptr<column>>
+  operator()(column_view const& all_keys,
+             column_view const& all_indices,
+             column_view const& new_keys,
+             offsets_pair const* d_offsets,
+             size_type const* d_map_to_keys,
+             cudaStream_t stream,
+             rmm::mr::device_memory_resource* mr)
+  {
+    auto keys_view     = column_device_view::create(all_keys, stream);
+    auto d_all_keys    = *keys_view;
+    auto indices_view  = column_device_view::create(all_indices, stream);
+    auto d_all_indices = *indices_view;
+
+    auto all_itr = thrust::make_transform_iterator(
+      thrust::make_counting_iterator<size_type>(0),
+      [d_all_keys, d_offsets, d_map_to_keys, d_all_indices] __device__(size_type idx) {
+        if (d_all_indices.is_null(idx)) return Element{};
+        size_type index =
+          d_all_indices.template element<int32_t>(idx) + d_offsets[d_map_to_keys[idx]].first;
+        return d_all_keys.template element<Element>(index);
+      });
+
+    auto new_keys_view = column_device_view::create(new_keys, stream);
+    auto d_new_keys    = *new_keys_view;
+    auto keys_itr      = thrust::make_transform_iterator(
+      thrust::make_counting_iterator<size_type>(0),
+      [d_new_keys] __device__(size_type idx) { return d_new_keys.template element<Element>(idx); });
+
+    // compute new indices values into output column
+    auto result = make_numeric_column(
+      data_type{type_id::INT32}, all_indices.size(), mask_state::UNALLOCATED, stream, mr);
+    auto d_result = result->mutable_view().data<int32_t>();
+    thrust::lower_bound(rmm::exec_policy(stream)->on(stream),
+                        keys_itr,
+                        keys_itr + new_keys.size(),
+                        all_itr,
+                        all_itr + all_indices.size(),
+                        d_result,
+                        thrust::less<Element>());
+    return result;
+  }
+
+  template <typename Element>
+  typename std::enable_if_t<!cudf::is_relationally_comparable<Element, Element>(),
+                            std::unique_ptr<column>>
+  operator()(column_view const&,
+             column_view const&,
+             column_view const&,
+             offsets_pair const*,
+             size_type const*,
+             cudaStream_t stream,
+             rmm::mr::device_memory_resource*)
+  {
+    CUDF_FAIL("list_view as keys for dictionary not supported");
+  }
+};
+
+}  // namespace
+
+std::unique_ptr<column> concatenate(std::vector<column_view> const& columns,
+                                    cudaStream_t stream,
+                                    rmm::mr::device_memory_resource* mr)
+{
+  if (columns.size() == 0) return make_empty_column(data_type{type_id::DICTIONARY32});
```

Comment:
Probably use `empty_like` for if/when we add other dictionary types.

Reply:
The `empty_like` requires a column_view parameter. 
https://docs.rapids.ai/api/libcudf/stable/group__column__copy.html#ga0e09ded998662935b453cf161c752e5c
There is no `column_view` to pass here.

Reply:
> There is no column_view to pass here.

What about the `std::vector<column_view> const& columns`? 

We made the decision that columns always need to preserve their hierarchies. A `DICTIONARY32` column without any children loses information because you no longer know what it is a dictionary _of_.

Reply:
Maybe I'm missing something, 
The if-clause is specifically checking `columns.size()==0` which means there are no `column_view's` in the `std::vector<column_view> const& column` variable and therefore I have no `column_view` in which to pass to `empty_like()`

Reply:
Ah, I mistook `columns.size()` for accessing a column, not a `column_view`. I think this should maybe be an exception then? I'm not sure what the behavior in other code paths is if you try and concatenate 0 columns.

Reply:
`cudf::concatenate` throws an exception for empty column
https://github.com/rapidsai/cudf/blob/b4874fb0affc22eae7235df37ba9150970565932/cpp/src/copying/concatenate.cu#L328
So changed this line to match that one